### DISCUSSION TITLE
Use of cuda::std::enable_if

### DISCUSSION THREAD

**Discussion by Artem-B** [NVIDIA/cccl#1693] [CUB] [answered]
Some of our code ran into a compilation error arrempting to use UCB with `CUB_NS_{PREFIX/POSTFIX/NS_QUALIFIER}` defined:
https://github.com/NVIDIA/cccl/blob/c910b166afd6fb90175df6c1f4aa792ed80a9214/cub/cub/agent/agent_three_way_partition.cuh#L96

```
In file included from third_party/py/dreamplace/ops/pin_pos/src/pin_pos_cuda_segment_kernel.cu:6:
In file included from ./third_party/py/dreamplace/ops/utility/src/utils_cub.cuh:18:
In file included from third_party/gpus/cccl/cub/cub/cub.cuh:67:
In file included from third_party/gpus/cccl/cub/cub/device/device_partition.cuh:51:
In file included from third_party/gpus/cccl/cub/cub/device/dispatch/dispatch_three_way_partition.cuh:40:
third_party/gpus/cccl/cub/cub/agent/agent_three_way_partition.cuh:87:50: error: no member named 'std' in namespace 'DreamPlace::cuda'; did you mean simply 'std'?
   87 | struct accumulator_pack_base_t<OffsetT, typename cuda::std::enable_if<sizeof(OffsetT) == 4>::type>
      |                                                  ^
third_party/crosstool/v18/stable/toolchain/bin/../include/c++/v1/__atomic/kill_dependency.h:18:1: note: 'std' declared here
   18 | _LIBCPP_BEGIN_NAMESPACE_STD
      | ^
third_party/crosstool/v18/stable/toolchain/bin/../include/c++/v1/__config:846:81: note: expanded from macro '_LIBCPP_BEGIN_NAMESPACE_STD'
  846 |                                       namespace _LIBCPP_TYPE_VISIBILITY_DEFAULT std {                                  \
      |                                                                                 ^
```

Using `::cuda::std::enable_if` appears to fix the problem, but I'm not sure if it's cutlass' bug or if something is messed up on the user's side. Hence a discussion and not a bug report yet.

@jrhemstad @gevtushenko -- any suggestions on what's the right way to deal with this? 

The code which triggers it is a rather old version of https://github.com/limbo018/DREAMPlace/tree/3.0.0, with `CUB_NS_QUALIFIER` added in https://github.com/limbo018/DREAMPlace/blob/3.0.0/dreamplace/ops/utility/src/utils_cub.cuh to make it work with cub 2.3.2.

**Comment by miscco** [comment, marked as answer]
That is a bug on our side. Any usage of namespace `cuda` should be fully qualified, aka `::cuda::std::`.

The reason is that `cuda` is not a reserved namespace as `std` is so users can and do happily nest it
